File: typo-tester-mariadb-java/src/java/testdb/CustomerRepoTest.java
package testdb;

import org.junit.Test;
import testdb.customer_status.CustomerStatusId;
import testdb.customer_status.CustomerStatusRepoImpl;
import testdb.customer_status.CustomerStatusRowUnsaved;
import testdb.customers.CustomersId;
import testdb.customers.CustomersRepoImpl;
import testdb.customers.CustomersRow;
import testdb.customers.CustomersRowUnsaved;
import testdb.customtypes.Defaulted.Provided;
import testdb.customtypes.Defaulted.UseDefault;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;

import static org.junit.Assert.*;

/**
 * Full repository test for customers - tests all CRUD operations and more.
 * Note: The database has seeded customer_status data including 'active', 'pending', 'suspended', 'closed'.
 * Tests use these existing statuses rather than inserting new ones.
 */
public class CustomerRepoTest {
    private final CustomersRepoImpl customersRepo = new CustomersRepoImpl();
    private final CustomerStatusRepoImpl customerStatusRepo = new CustomerStatusRepoImpl();

    // Reference to existing seeded status in the database
    private static final CustomerStatusId ACTIVE_STATUS = new CustomerStatusId("active");
    private static final CustomerStatusId PENDING_STATUS = new CustomerStatusId("pending");

    @Test
    public void testSelectAll() {
        WithConnection.run(c -> {
            // Use existing seeded status 'active'
            // Insert multiple customers using TestInsert
            var testInsert = new TestInsert(new Random(0));
            var customer1 = testInsert.Customers(
                "password123".getBytes(),
                "john@example.com",
                "John",
                "Doe",
                new UseDefault<>(),
                new Provided<>(ACTIVE_STATUS),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                c
            );
            var customer2 = testInsert.Customers(
                "password456".getBytes(),
                "jane@example.com",
                "Jane",
                "Smith",
                new UseDefault<>(),
                new Provided<>(ACTIVE_STATUS),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>(),
                c
            );

            var all = customersRepo.selectAll(c);
            assertEquals(2, all.size());
        });
    }

    @Test
    public void testSelectById() {
        WithConnection.run(c -> {
            // Uses default 'pending' status from database
            var unsaved = new CustomersRowUnsaved("test@example.com", "hash".getBytes(), "Test", "User");
            var inserted = customersRepo.insert(unsaved, c);

            var selected = customersRepo.selectById(inserted.customerId(), c);
            assertTrue(selected.isPresent());
            assertEquals("test@example.com", selected.get().email());
            assertEquals("Test", selected.get().firstName());
            assertEquals("User", selected.get().lastName());
        });
    }

    @Test
    public void testSelectByIds() {
        WithConnection.run(c -> {
            // Use existing seeded status
            var customer1 = customersRepo.insert(
                new CustomersRowUnsaved("c1@example.com", "hash1".getBytes(), "Customer", "One"),
                c
            );
            var customer2 = customersRepo.insert(
                new CustomersRowUnsaved("c2@example.com", "hash2".getBytes(), "Customer", "Two"),
                c
            );
            var customer3 = customersRepo.insert(
                new CustomersRowUnsaved("c3@example.com", "hash3".getBytes(), "Customer", "Three"),
                c
            );

            var ids = new CustomersId[]{customer1.customerId(), customer3.customerId()};
            var selected = customersRepo.selectByIds(ids, c);

            assertEquals(2, selected.size());
        });
    }

    @Test
    public void testSelectByIdsTracked() {
        WithConnection.run(c -> {
            // Use existing seeded status
            var customer1 = customersRepo.insert(
                new CustomersRowUnsaved("c1@example.com", "hash1".getBytes(), "Customer", "One"),
                c
            );
            var customer2 = customersRepo.insert(
                new CustomersRowUnsaved("c2@example.com", "hash2".getBytes(), "Customer", "Two"),
                c
            );

            var ids = new CustomersId[]{customer1.customerId(), customer2.customerId()};
            Map<CustomersId, CustomersRow> tracked = customersRepo.selectByIdsTracked(ids, c);

            assertEquals(2, tracked.size());
            assertEquals("Customer", tracked.get(customer1.customerId()).firstName());
            assertEquals("Two", tracked.get(customer2.customerId()).lastName());
        });
    }

    @Test
    public void testSelectByUniqueEmail() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var inserted = customersRepo.insert(
                new CustomersRowUnsaved("unique@example.com", "hash".getBytes(), "Unique", "User"),
                c
            );

            var selected = customersRepo.selectByUniqueEmail("unique@example.com", c);
            assertTrue(selected.isPresent());
            assertEquals(inserted.customerId(), selected.get().customerId());

            var notFound = customersRepo.selectByUniqueEmail("nonexistent@example.com", c);
            assertFalse(notFound.isPresent());
        });
    }

    @Test
    public void testInsertRow() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var unsaved = new CustomersRowUnsaved("insert@example.com", "password".getBytes(), "Insert", "Test");
            var inserted = customersRepo.insert(unsaved, c);

            assertNotNull(inserted.customerId());
            assertEquals("insert@example.com", inserted.email());
            assertEquals("Insert", inserted.firstName());
            assertEquals("Test", inserted.lastName());
            assertEquals("pending", inserted.status().value()); // default status
            assertEquals("bronze", inserted.tier()); // default tier
        });
    }

    @Test
    public void testInsertRowWithAllFields() {
        WithConnection.run(c -> {
            // Use existing seeded 'suspended' status
            var unsaved = new CustomersRowUnsaved(
                "premium@example.com",
                "securepass".getBytes(),
                "Premium",
                "Customer",
                new Provided<>(Optional.of("+1234567890")),
                new Provided<>(new CustomerStatusId("suspended")),
                new Provided<>("gold"),
                new Provided<>(Optional.of("{\"lang\": \"en\"}")),
                new UseDefault<>(),
                new Provided<>(Optional.of("Important customer")),
                new UseDefault<>(),
                new UseDefault<>(),
                new UseDefault<>()
            );

            var inserted = customersRepo.insert(unsaved, c);

            assertEquals("premium@example.com", inserted.email());
            assertEquals(Optional.of("+1234567890"), inserted.phone());
            assertEquals("suspended", inserted.status().value());
            assertEquals("gold", inserted.tier());
            assertEquals(Optional.of("{\"lang\": \"en\"}"), inserted.preferences());
            assertEquals(Optional.of("Important customer"), inserted.notes());
        });
    }

    @Test
    public void testUpdate() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var inserted = customersRepo.insert(
                new CustomersRowUnsaved("update@example.com", "hash".getBytes(), "Before", "Update"),
                c
            );

            var updated = inserted
                .withFirstName("After")
                .withLastName("Changed")
                .withTier("silver");

            var success = customersRepo.update(updated, c);
            assertTrue(success);

            var selected = customersRepo.selectById(inserted.customerId(), c);
            assertTrue(selected.isPresent());
            assertEquals("After", selected.get().firstName());
            assertEquals("Changed", selected.get().lastName());
            assertEquals("silver", selected.get().tier());
        });
    }

    @Test
    public void testUpsertInsert() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            // First insert a row to get a valid customerId
            var initial = customersRepo.insert(
                new CustomersRowUnsaved("upsert@example.com", "hash".getBytes(), "Upsert", "Test"),
                c
            );

            // Delete it
            customersRepo.deleteById(initial.customerId(), c);

            // Now upsert with a new row (using same id which should work for new insert)
            var row = new CustomersRow(
                initial.customerId(),
                "upsert2@example.com",
                "newhash".getBytes(),
                "New",
                "Upsert",
                Optional.empty(),
                new CustomerStatusId("pending"),
                "bronze",
                Optional.empty(),
                Optional.empty(),
                Optional.empty(),
                initial.createdAt(),
                initial.updatedAt(),
                Optional.empty()
            );

            var upserted = customersRepo.upsert(row, c);
            assertEquals("upsert2@example.com", upserted.email());
        });
    }

    @Test
    public void testUpsertUpdate() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var inserted = customersRepo.insert(
                new CustomersRowUnsaved("upsert.update@example.com", "hash".getBytes(), "Before", "Upsert"),
                c
            );

            var modified = inserted
                .withFirstName("After")
                .withTier("gold");

            var upserted = customersRepo.upsert(modified, c);
            assertEquals("After", upserted.firstName());
            assertEquals("gold", upserted.tier());

            // Verify it's the same row
            assertEquals(inserted.customerId(), upserted.customerId());
        });
    }

    @Test
    public void testUpsertBatch() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var c1 = customersRepo.insert(
                new CustomersRowUnsaved("batch1@example.com", "hash1".getBytes(), "Batch", "One"),
                c
            );
            var c2 = customersRepo.insert(
                new CustomersRowUnsaved("batch2@example.com", "hash2".getBytes(), "Batch", "Two"),
                c
            );

            // Modify both
            var m1 = c1.withFirstName("Modified1");
            var m2 = c2.withFirstName("Modified2");

            var upserted = customersRepo.upsertBatch(List.of(m1, m2).iterator(), c);
            assertEquals(2, upserted.size());

            // Verify changes
            var s1 = customersRepo.selectById(c1.customerId(), c);
            var s2 = customersRepo.selectById(c2.customerId(), c);
            assertEquals("Modified1", s1.get().firstName());
            assertEquals("Modified2", s2.get().firstName());
        });
    }

    @Test
    public void testDeleteById() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var inserted = customersRepo.insert(
                new CustomersRowUnsaved("delete@example.com", "hash".getBytes(), "Delete", "Me"),
                c
            );

            var deleted = customersRepo.deleteById(inserted.customerId(), c);
            assertTrue(deleted);

            var selected = customersRepo.selectById(inserted.customerId(), c);
            assertFalse(selected.isPresent());
        });
    }

    @Test
    public void testDeleteByIds() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var c1 = customersRepo.insert(
                new CustomersRowUnsaved("del1@example.com", "hash1".getBytes(), "Del", "One"),
                c
            );
            var c2 = customersRepo.insert(
                new CustomersRowUnsaved("del2@example.com", "hash2".getBytes(), "Del", "Two"),
                c
            );
            var c3 = customersRepo.insert(
                new CustomersRowUnsaved("del3@example.com", "hash3".getBytes(), "Del", "Three"),
                c
            );

            var ids = new CustomersId[]{c1.customerId(), c3.customerId()};
            var count = customersRepo.deleteByIds(ids, c);
            assertEquals(Integer.valueOf(2), count);

            var remaining = customersRepo.selectAll(c);
            assertEquals(1, remaining.size());
            assertEquals(c2.customerId(), remaining.get(0).customerId());
        });
    }

    @Test
    public void testDSLSelect() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            customersRepo.insert(
                new CustomersRowUnsaved("dsl1@example.com", "hash1".getBytes(), "Alice", "Smith"),
                c
            );
            customersRepo.insert(
                new CustomersRowUnsaved("dsl2@example.com", "hash2".getBytes(), "Bob", "Jones"),
                c
            );
            customersRepo.insert(
                new CustomersRowUnsaved("dsl3@example.com", "hash3".getBytes(), "Alice", "Brown"),
                c
            );

            // Test select with where clause
            var alices = customersRepo.select()
                .where(f -> f.firstName().isEqual("Alice"))
                .toList(c);
            assertEquals(2, alices.size());

            // Test select with multiple conditions
            var specificAlice = customersRepo.select()
                .where(f -> f.firstName().isEqual("Alice"))
                .where(f -> f.lastName().isEqual("Smith"))
                .toList(c);
            assertEquals(1, specificAlice.size());
            assertEquals("Alice", specificAlice.get(0).firstName());
            assertEquals("Smith", specificAlice.get(0).lastName());
        });
    }

    @Test
    public void testDSLUpdate() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            var customer = customersRepo.insert(
                new CustomersRowUnsaved("dsl.update@example.com", "hash".getBytes(), "Before", "Update"),
                c
            );

            customersRepo.update()
                .setValue(f -> f.firstName(), "After")
                .setValue(f -> f.tier(), "platinum")
                .where(f -> f.customerId().isEqual(customer.customerId()))
                .execute(c);

            var updated = customersRepo.selectById(customer.customerId(), c);
            assertTrue(updated.isPresent());
            assertEquals("After", updated.get().firstName());
            assertEquals("platinum", updated.get().tier());
        });
    }

    @Test
    public void testDSLDelete() {
        WithConnection.run(c -> {
            // Use existing seeded 'pending' status (default)
            customersRepo.insert(
                new CustomersRowUnsaved("dsl.del1@example.com", "hash1".getBytes(), "ToDelete", "One"),
                c
            );
            customersRepo.insert(
                new CustomersRowUnsaved("dsl.del2@example.com", "hash2".getBytes(), "ToDelete", "Two"),
                c
            );
            customersRepo.insert(
                new CustomersRowUnsaved("dsl.keep@example.com", "hash3".getBytes(), "ToKeep", "One"),
                c
            );

            customersRepo.delete()
                .where(f -> f.firstName().isEqual("ToDelete"))
                .execute(c);

            var remaining = customersRepo.selectAll(c);
            assertEquals(1, remaining.size());
            assertEquals("ToKeep", remaining.get(0).firstName());
        });
    }
}

File: typo-tester-mariadb-kotlin/src/kotlin/testdb/CustomerRepoTest.kt
package testdb

import org.junit.Assert.*
import org.junit.Test
import testdb.customer_status.CustomerStatusId
import testdb.customer_status.CustomerStatusRepoImpl
import testdb.customers.CustomersId
import testdb.customers.CustomersRepoImpl
import testdb.customers.CustomersRow
import testdb.customers.CustomersRowUnsaved
import testdb.customtypes.Defaulted.Provided
import testdb.customtypes.Defaulted.UseDefault

/**
 * Full repository test for customers - tests all CRUD operations and more.
 * Note: The database has seeded customer_status data including 'active', 'pending', 'suspended', 'closed'.
 * Tests use these existing statuses rather than inserting new ones.
 */
class CustomerRepoTest {
    private val customersRepo = CustomersRepoImpl()
    private val customerStatusRepo = CustomerStatusRepoImpl()

    companion object {
        private val ACTIVE_STATUS = CustomerStatusId("active")
        private val PENDING_STATUS = CustomerStatusId("pending")
    }

    @Test
    fun testSelectAll() {
        WithConnection.run { c ->
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Customer",
                    lastName = "One"
                ),
                c
            )
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Customer",
                    lastName = "Two"
                ),
                c
            )

            val all = customersRepo.selectAll(c)
            assertEquals(2, all.size)
        }
    }

    @Test
    fun testSelectById() {
        WithConnection.run { c ->
            val unsaved = CustomersRowUnsaved(
                email = "test@example.com",
                passwordHash = "hash".toByteArray(),
                firstName = "Test",
                lastName = "User"
            )
            val inserted = customersRepo.insert(unsaved, c)

            val selected = customersRepo.selectById(inserted.customerId, c)
            assertNotNull(selected)
            assertEquals("test@example.com", selected!!.email)
            assertEquals("Test", selected.firstName)
            assertEquals("User", selected.lastName)
        }
    }

    @Test
    fun testSelectByIds() {
        WithConnection.run { c ->
            val customer1 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Customer",
                    lastName = "One"
                ),
                c
            )
            val customer2 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Customer",
                    lastName = "Two"
                ),
                c
            )
            val customer3 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c3@example.com",
                    passwordHash = "hash3".toByteArray(),
                    firstName = "Customer",
                    lastName = "Three"
                ),
                c
            )

            val ids = arrayOf(customer1.customerId, customer3.customerId)
            val selected = customersRepo.selectByIds(ids, c)

            assertEquals(2, selected.size)
        }
    }

    @Test
    fun testSelectByIdsTracked() {
        WithConnection.run { c ->
            val customer1 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Customer",
                    lastName = "One"
                ),
                c
            )
            val customer2 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "c2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Customer",
                    lastName = "Two"
                ),
                c
            )

            val ids = arrayOf(customer1.customerId, customer2.customerId)
            val tracked = customersRepo.selectByIdsTracked(ids, c)

            assertEquals(2, tracked.size)
            assertEquals("Customer", tracked[customer1.customerId]?.firstName)
            assertEquals("Two", tracked[customer2.customerId]?.lastName)
        }
    }

    @Test
    fun testSelectByUniqueEmail() {
        WithConnection.run { c ->
            val inserted = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "unique@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Unique",
                    lastName = "User"
                ),
                c
            )

            val selected = customersRepo.selectByUniqueEmail("unique@example.com", c)
            assertNotNull(selected)
            assertEquals(inserted.customerId, selected!!.customerId)

            val notFound = customersRepo.selectByUniqueEmail("nonexistent@example.com", c)
            assertNull(notFound)
        }
    }

    @Test
    fun testInsertRow() {
        WithConnection.run { c ->
            val unsaved = CustomersRowUnsaved(
                email = "insert@example.com",
                passwordHash = "password".toByteArray(),
                firstName = "Insert",
                lastName = "Test"
            )
            val inserted = customersRepo.insert(unsaved, c)

            assertNotNull(inserted.customerId)
            assertEquals("insert@example.com", inserted.email)
            assertEquals("Insert", inserted.firstName)
            assertEquals("Test", inserted.lastName)
            assertEquals("pending", inserted.status.value)
            assertEquals("bronze", inserted.tier)
        }
    }

    @Test
    fun testInsertRowWithAllFields() {
        WithConnection.run { c ->
            val unsaved = CustomersRowUnsaved(
                email = "premium@example.com",
                passwordHash = "securepass".toByteArray(),
                firstName = "Premium",
                lastName = "Customer",
                phone = Provided("+1234567890"),
                status = Provided(CustomerStatusId("suspended")),
                tier = Provided("gold"),
                preferences = Provided("{\"lang\": \"en\"}"),
                notes = Provided("Important customer")
            )

            val inserted = customersRepo.insert(unsaved, c)

            assertEquals("premium@example.com", inserted.email)
            assertEquals("+1234567890", inserted.phone)
            assertEquals("suspended", inserted.status.value)
            assertEquals("gold", inserted.tier)
            assertEquals("{\"lang\": \"en\"}", inserted.preferences)
            assertEquals("Important customer", inserted.notes)
        }
    }

    @Test
    fun testUpdate() {
        WithConnection.run { c ->
            val inserted = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "update@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Before",
                    lastName = "Update"
                ),
                c
            )

            val updated = inserted.copy(
                firstName = "After",
                lastName = "Changed",
                tier = "silver"
            )

            val success = customersRepo.update(updated, c)
            assertTrue(success)

            val selected = customersRepo.selectById(inserted.customerId, c)
            assertNotNull(selected)
            assertEquals("After", selected!!.firstName)
            assertEquals("Changed", selected.lastName)
            assertEquals("silver", selected.tier)
        }
    }

    @Test
    fun testUpsertInsert() {
        WithConnection.run { c ->
            val initial = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "upsert@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Upsert",
                    lastName = "Test"
                ),
                c
            )

            customersRepo.deleteById(initial.customerId, c)

            val row = CustomersRow(
                customerId = initial.customerId,
                email = "upsert2@example.com",
                passwordHash = "newhash".toByteArray(),
                firstName = "New",
                lastName = "Upsert",
                phone = null,
                status = CustomerStatusId("pending"),
                tier = "bronze",
                preferences = null,
                marketingFlags = null,
                notes = null,
                createdAt = initial.createdAt,
                updatedAt = initial.updatedAt,
                lastLoginAt = null
            )

            val upserted = customersRepo.upsert(row, c)
            assertEquals("upsert2@example.com", upserted.email)
        }
    }

    @Test
    fun testUpsertUpdate() {
        WithConnection.run { c ->
            val inserted = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "upsert.update@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Before",
                    lastName = "Upsert"
                ),
                c
            )

            val modified = inserted.copy(
                firstName = "After",
                tier = "gold"
            )

            val upserted = customersRepo.upsert(modified, c)
            assertEquals("After", upserted.firstName)
            assertEquals("gold", upserted.tier)
            assertEquals(inserted.customerId, upserted.customerId)
        }
    }

    @Test
    fun testUpsertBatch() {
        WithConnection.run { c ->
            val c1 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "batch1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Batch",
                    lastName = "One"
                ),
                c
            )
            val c2 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "batch2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Batch",
                    lastName = "Two"
                ),
                c
            )

            val m1 = c1.copy(firstName = "Modified1")
            val m2 = c2.copy(firstName = "Modified2")

            val upserted = customersRepo.upsertBatch(mutableListOf(m1, m2).iterator(), c)
            assertEquals(2, upserted.size)

            val s1 = customersRepo.selectById(c1.customerId, c)
            val s2 = customersRepo.selectById(c2.customerId, c)
            assertEquals("Modified1", s1!!.firstName)
            assertEquals("Modified2", s2!!.firstName)
        }
    }

    @Test
    fun testDeleteById() {
        WithConnection.run { c ->
            val inserted = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "delete@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Delete",
                    lastName = "Me"
                ),
                c
            )

            val deleted = customersRepo.deleteById(inserted.customerId, c)
            assertTrue(deleted)

            val selected = customersRepo.selectById(inserted.customerId, c)
            assertNull(selected)
        }
    }

    @Test
    fun testDeleteByIds() {
        WithConnection.run { c ->
            val c1 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "del1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Del",
                    lastName = "One"
                ),
                c
            )
            val c2 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "del2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Del",
                    lastName = "Two"
                ),
                c
            )
            val c3 = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "del3@example.com",
                    passwordHash = "hash3".toByteArray(),
                    firstName = "Del",
                    lastName = "Three"
                ),
                c
            )

            val ids = arrayOf(c1.customerId, c3.customerId)
            val count = customersRepo.deleteByIds(ids, c)
            assertEquals(2, count)

            val remaining = customersRepo.selectAll(c)
            assertEquals(1, remaining.size)
            assertEquals(c2.customerId, remaining[0].customerId)
        }
    }

    @Test
    fun testDSLSelect() {
        WithConnection.run { c ->
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "Alice",
                    lastName = "Smith"
                ),
                c
            )
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "Bob",
                    lastName = "Jones"
                ),
                c
            )
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl3@example.com",
                    passwordHash = "hash3".toByteArray(),
                    firstName = "Alice",
                    lastName = "Brown"
                ),
                c
            )

            val alices = customersRepo.select()
                .where { f -> f.firstName().isEqual("Alice") }
                .toList(c)
            assertEquals(2, alices.size)

            val specificAlice = customersRepo.select()
                .where { f -> f.firstName().isEqual("Alice") }
                .where { f -> f.lastName().isEqual("Smith") }
                .toList(c)
            assertEquals(1, specificAlice.size)
            assertEquals("Alice", specificAlice[0].firstName)
            assertEquals("Smith", specificAlice[0].lastName)
        }
    }

    @Test
    fun testDSLUpdate() {
        WithConnection.run { c ->
            val customer = customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl.update@example.com",
                    passwordHash = "hash".toByteArray(),
                    firstName = "Before",
                    lastName = "Update"
                ),
                c
            )

            customersRepo.update()
                .setValue({ f -> f.firstName() }, "After")
                .setValue({ f -> f.tier() }, "platinum")
                .where { f -> f.customerId().isEqual(customer.customerId) }
                .execute(c)

            val updated = customersRepo.selectById(customer.customerId, c)
            assertNotNull(updated)
            assertEquals("After", updated!!.firstName)
            assertEquals("platinum", updated.tier)
        }
    }

    @Test
    fun testDSLDelete() {
        WithConnection.run { c ->
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl.del1@example.com",
                    passwordHash = "hash1".toByteArray(),
                    firstName = "ToDelete",
                    lastName = "One"
                ),
                c
            )
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl.del2@example.com",
                    passwordHash = "hash2".toByteArray(),
                    firstName = "ToDelete",
                    lastName = "Two"
                ),
                c
            )
            customersRepo.insert(
                CustomersRowUnsaved(
                    email = "dsl.keep@example.com",
                    passwordHash = "hash3".toByteArray(),
                    firstName = "ToKeep",
                    lastName = "One"
                ),
                c
            )

            customersRepo.delete()
                .where { f -> f.firstName().isEqual("ToDelete") }
                .execute(c)

            val remaining = customersRepo.selectAll(c)
            assertEquals(1, remaining.size)
            assertEquals("ToKeep", remaining[0].firstName)
        }
    }
}
